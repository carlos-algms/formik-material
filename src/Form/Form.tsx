import { Form as FormikForm, Formik, FormikErrors, FormikHelpers, FormikProps } from 'formik';
import React from 'react';
import type PropsBase from '../shared/PropsBase';

export type OnSubmitFormCallback<T = any, R = any> = {
  (values: T, formikHelpers: FormikHelpers<T>): R | Promise<R>;
};

export type FormProps<T> = PropsBase<FormikProps<T>> & {
  initialValues?: T;
  onSubmitForm: OnSubmitFormCallback<T>;
  validationSchema?: any | (() => any);
  validate?: (values: T) => FormikErrors<T> | Promise<FormikErrors<T>>;
};

/**
 * Form Component
 */
export default function Form<T>({
  children,
  initialValues,
  onSubmitForm,
  validationSchema,
  validate,
  className,
}: FormProps<T>) {
  if (!initialValues) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn('No initial values provided to the <Form> Component');
    }
    return null;
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmitForm}
      validationSchema={validationSchema}
      validate={validate}
    >
      {(formikProps) => {
        return (
          <FormikForm noValidate autoComplete="off" className={className}>
            {typeof children === 'function' ? children(formikProps) : children}
          </FormikForm>
        );
      }}
    </Formik>
  );
}
