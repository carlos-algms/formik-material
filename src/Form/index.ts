export { default } from './Form';
export type { FormProps, OnSubmitFormCallback } from './Form';
