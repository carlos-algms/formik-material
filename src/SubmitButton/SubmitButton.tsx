import Button from '@material-ui/core/Button';
import { useFormikContext } from 'formik';
import React, { memo } from 'react';
import type PropsBase from '../shared/PropsBase';

function SubmitButton({ children, className }: PropsBase) {
	const { isSubmitting } = useFormikContext();

	return (
		<div className={className}>
			<Button variant="contained" color="primary" type="submit" disabled={isSubmitting}>
				{children}
			</Button>
		</div>
	);
}

export default memo(SubmitButton);
