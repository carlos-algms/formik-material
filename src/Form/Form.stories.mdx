import { action } from '@storybook/addon-actions';
import { Meta, Canvas, Story, Props } from '@storybook/addon-docs/blocks';
import Form from './Form';
import SubmitButton from '../SubmitButton';
import useFormSubmittedValues from '../useFormSubmittedValues';
import TextField from '../TextField';

export const initialValues = { name: 'initial name' };

<Meta title="Form Component" component={Form} />

## Basic form implementation

<Canvas>
  <Story name="basic">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          <TextField name="name" label="Name: " />
          <p>
            <SubmitButton>Submit</SubmitButton>
          </p>
        </Form>
      );
    }}
  </Story>
</Canvas>

## Forms should be able to render child functions

<Canvas>
  <Story name="render child function">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          {() => (
            <>
              <TextField name="name" label="Name: " />
              <p>
                <SubmitButton>Submit</SubmitButton>
              </p>
            </>
          )}
        </Form>
      );
    }}
  </Story>
</Canvas>
