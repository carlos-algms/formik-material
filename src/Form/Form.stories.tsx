import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';
import SubmitButton from '../SubmitButton';
import TextField from '../TextField';
import useFormSubmittedValues from '../useFormSubmittedValues';
import Form from './Form';

const initialValues = { name: 'initial name' };

const meta: Meta<typeof Form> = {
  title: 'Form Component',
  component: Form,
};

export default meta;

type Story = StoryObj<typeof Form>;

export const Basic: Story = {
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        <TextField name="name" label="Name: " />
        <p>
          <SubmitButton>Submit</SubmitButton>
        </p>
      </Form>
    );
  },
};

export const ChildFn: Story = {
  name: 'Render children as function',
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        {() => (
          <>
            <TextField name="name" label="Name: " />
            <p>
              <SubmitButton>Submit</SubmitButton>
            </p>
          </>
        )}
      </Form>
    );
  },
};
