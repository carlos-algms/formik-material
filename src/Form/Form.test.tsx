import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import SubmitButton from '../SubmitButton';
import { delay } from '../tests/helpers';
import Form from './Form';

test('should render a form', () => {
  const submitMock = jest.fn();
  const { container } = render(
    <Form className="my-form" onSubmitForm={submitMock} initialValues={{ name: 'foo' }} />,
  );

  expect(container).toMatchSnapshot();
});

test('should render children when a function is passed', () => {
  const submitMock = jest.fn();
  const initialValues = { name: 'foo' };
  let childrenProps: any;

  render(
    <Form onSubmitForm={submitMock} initialValues={initialValues}>
      {(props) => {
        childrenProps = props;
        return <SubmitButton>Send</SubmitButton>;
      }}
    </Form>,
  );

  const submitButton = screen.getByRole('button');
  expect(submitButton).toBeInTheDocument();

  expect(childrenProps.values).toEqual(initialValues);
  expect(childrenProps.isValid).toBeTruthy();
});

test('should not call onSubmit on fail validation', async () => {
  const validateMock = jest.fn();
  const submitMock = jest.fn();

  render(
    <Form validate={validateMock} onSubmitForm={submitMock} initialValues={{ name: 'foo' }}>
      <SubmitButton>Send</SubmitButton>
    </Form>,
  );

  let submitButton = screen.getByRole('button');

  validateMock.mockReturnValue({
    name: 'not valid',
  });

  await act(async () => {
    fireEvent.click(submitButton);
    await delay();
  });

  await delay(() => {
    expect(validateMock).toHaveBeenCalledTimes(1);
    expect(submitMock).not.toHaveBeenCalled();
  });

  validateMock.mockReturnValue({});
  await act(async () => {
    fireEvent.click(submitButton);
    await delay();
  });

  await delay(() => {
    expect(validateMock).toHaveBeenCalledTimes(2);
    expect(submitMock).toHaveBeenCalled();
  });
});
