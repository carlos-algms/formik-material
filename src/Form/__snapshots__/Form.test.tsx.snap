// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a form 1`] = `
<div>
  <form
    action="#"
    autocomplete="off"
    class="my-form"
    novalidate=""
  />
</div>
`;
